src/raft/cluster.rs

```diff
@@ -262,19 +283,32 @@ fn server_map(servers: Vec<Server>) -> HashMap<String, Server> {
     servers.into_iter().map(|s| (key(&s), s.clone())).collect()
 }
 
-fn is_quorum_among(votes: &HashSet<String>, members: &HashMap<String, Server>) -> bool {
+// Returns whether the supplied votes constitute a quorum (i.e. majority) among the supplied
+// members. Callers mustn't pass the same string in both "aye" and "nay".
+fn has_quorum_among(
+    members: &HashSet<String>,
+    aye: &HashSet<String>,
+    nay: &HashSet<String>,
+) -> QuorumResult {
     if members.is_empty() {
-        return true;
+        return QuorumResult::Yes;
     }
 
-    // Count the number present in the members.
-    let mut count = 0;
-    for key in votes {
-        if members.contains_key(key.as_str()) {
-            count = count + 1;
-        }
+    let mem_count = members.len();
+    let aye_count = aye.iter().filter(|&s| members.contains(s.as_str())).count();
+    let nay_count = nay.iter().filter(|&s| members.contains(s.as_str())).count();
+
+    let remaining = mem_count - (aye_count + nay_count);
+    let can_yes_win = 2 * (aye_count + remaining) > mem_count;
+    let can_no_win = 2 * (nay_count + remaining) > mem_count;
+
+    if can_yes_win && !can_no_win {
```

Comment:
Need to check if this behaves correctly...

Do we declare victory before receiving all the necessary yes votes? Do we actually need both && conditions or could we, say, declare victory when !can_no_win ?